### DISCUSSION TITLE
Dev Version of MIPModel?

### DISCUSSION THREAD

**Discussion by datadrivensupplychain** [dirkschumacher/ompr#428] [Q&A]
In another thread, I read that the dev version of `MIPModel` has incorporated the speed-up capabilities of MILPModel.  I am starting a new project and just installed the dev versions of `ompr` (1.0.2.9000) and `ompr.roi` (1.0.0.9000).  I don't have that big of a model. Defining it as a `MILPModel`, it builds in less than a second.  As a 1MIPModel1 - it's at a couple minutes and not even halfway through the build!  Is there something I'm missing, or does `MIPModel` not yet have the speed of `MILPModel`?

**Comment by sbmack** [comment]
Right, an orders of magnitude performance difference does not appear to make sense.  Can you post both the MIPModel and MILPModel formulations?

The latest release notes from Dirk can be found in this discussion item from January 2022:
https://github.com/dirkschumacher/ompr/discussions/397#discussion-3835462

The version 1.0.2.9000 release date is 2022-03-14, so after his discussion update.

Note that since this is a side project for Dirk, he has not conducted formal performance testing of updated MIPModel vs. MILPModel.

**Reply by datadrivensupplychain** [reply]
No difference in model formulations besides the instantiation function

```
my_model <- ompr::MILPModel() %>%
  #add decision variables: flow from producer to ports, by SKU. continuous, zero lowerbound
  add_variable(producer_port_flow[producerindex,portindex,skuindex],
               producerindex=1:producer_count,portindex=1:port_count,skuindex=1:sku_count,type='continuous',lb=0) %>%
  #flow from ports to 3pl
  add_variable(port_3pl_flow[portindex,threeplindex,skuindex],
              portindex=1:port_count,threeplindex=1:threepl_count, skuindex=1:sku_count,type='continuous',lb=0) %>%
  #flow from ports directly to customer
  add_variable(port_customer_flow[portindex,customerindex,skuindex],
               portindex=1:port_count,customerindex=1:customer_count, skuindex=1:sku_count,type='continuous',lb=0) %>%
  #flow from 3PL to customer
  add_variable(threepl_customer_flow[threeplindex,customerindex,skuindex],
               threeplindex=1:threepl_count,customerindex=1:customer_count, skuindex=1:sku_count,type='continuous',lb=0) %>%
  #flow from ports to tollers
  add_variable(port_toller_flow[portindex,tollerindex,skuindex],
               portindex=1:port_count,tollerindex=1:toller_count, skuindex=1:sku_count,type='continuous',lb=0) %>%
  #flow from 3PL to tollers
  add_variable(threepl_toller_flow[threeplindex,tollerindex,skuindex],
               threeplindex=1:threepl_count,tollerindex=1:toller_count, skuindex=1:sku_count,type='continuous',lb=0) %>%
  #flow from toller to customer
  add_variable(toller_customer_flow[tollerindex,customerindex,skuindex],
               tollerindex=1:toller_count,customerindex=1:customer_count, skuindex=1:sku_count,type='continuous',lb=0) %>%
  #open/close binary decision variables for each building, by type
  add_variable(port_openclose_binary[portindex],portindex=1:port_count,type='binary') %>%
  add_variable(threepl_openclose_binary[threeplindex],threeplindex=1:threepl_count,type='binary') %>%
  add_variable(toller_openclose_binary[tollerindex],tollerindex=1:toller_count,type='binary')

#flow constraints ----
#ports flow constraints
my_model %<>% add_constraint(
  #LHS into ports from international producers
  sum_expr(producer_port_flow[producerindex,portindex,skuindex],producerindex=1:producer_count) ==
    #RHS out of ports to 3PL ,  customer , Toller
    sum_expr(port_3pl_flow[portindex,threeplindex,skuindex],threeplindex=1:threepl_count)+
    sum_expr(port_customer_flow[portindex,customerindex,skuindex],customerindex=1:customer_count) +
    sum_expr(port_toller_flow[portindex,tollerindex,skuindex],tollerindex=1:toller_count),
  portindex=1:port_count,skuindex=1:sku_count)

#3pl flow constraints
my_model %<>% add_constraint(
  #LHS into 3PL from ports
  sum_expr(port_3pl_flow[portindex,threeplindex,skuindex],portindex=1:port_count) ==
    #RHS out of 3PL to Toller and customer
    sum_expr(threepl_toller_flow[threeplindex,tollerindex,skuindex],tollerindex=1:toller_count)+
    sum_expr(threepl_customer_flow[threeplindex,customerindex,skuindex],customerindex=1:customer_count) ,
  threeplindex=1:threepl_count,skuindex=1:sku_count)
```

**Reply by datadrivensupplychain** [reply]
Also if there's a way to support the project so it's not just Dirk's side project, I'd be interested in learning more.  I've certainly gained my share of commercial value from this package.

**Reply by sbmack** [reply]
> No difference in model formulations besides the instantiation function

For grins I created dummy values for your model parameters and ran both the MILPModel and MIPModel cases:
```
# n is to scale the test
n <- 1
producer_count <- 5 * n
port_count <- 3 * n
sku_count <- 10 * n
threepl_count <- 3 * n
customer_count <- 4 * n
toller_count <- 20 * n
```
If you pass those values to `my_model <- MILPModel`, the object generated is a list of length 5, Type milp_model and size 310KB.

If you pass the same values to `my_model2 <- MIPModel`, the object generated is a list of length 7, Type linear_optimization_model and size 10.9MB so over 35X larger.

So 2 different objects are being generated.  View both list objects in the environment to visually compare the summaries.  However if you enter my_model and my_model2 at the console the ompr model structures appear to be notionally the same.  However `identical(my_model, my_model2)` of course returns `FALSE` since the R list structures are different.

Also note that if you just generate the variables of each model you will see that the size inflation happens at variable creation.

So that's the diagnosis based on my intellectual curiosity.  Perhaps Dirk can offer guidance from here.  Good Luck.  SteveM

**Reply by datadrivensupplychain** [reply]
Huh I wonder why the variable size are so fundamentally different.  I've been using MILPModel for my large models in previous projects because they build so quickly, which can matter a lot when I'm talking about 1M+ decision variables.  In reviewing the results at a decision-variable level, nothing looks odd.  @dirkschumacher any background / guidance is appreciated.

**Reply by sbmack** [reply]
I'm not saying which one is correct.  But note that constraint matrix generation with `MILPModel` often/usually requires the use of the `colwise` function.  So is your `MILPModel` version without `colwise` really giving you what you want?

With other model management platforms I often write out a reduced size model in standard form so that I can visually inspect it to ensure it's giving me what I want.  Unfortunately the ROI write function generates a very primitive representation of a model in standard form so has limited utility.  That's another shortfall for using ompr now as a real-use platform.  BTW, I agree that ompr is a great start to a complete model management platform.

**Reply by datadrivensupplychain** [reply]
Fair point, and I do use `colwise` for objective functions and scalar coefficients.  The constraints in the model I posted above is simply defining relationships between decision variables.  In supply chain modeling, a "flow constraint" simply means that the volume going into a facility, must equal the volume leaving the facility.  Once I get to the point of adding an objective function, I'll use `colwise` (and functions I've written that use `colwise` , allowing a modeler to directly multiply a matrix of decision variables by a scalar matrix of equal dimensions)
